### PR TITLE
Cog support cesium only 1

### PR DESCRIPTION
**Pull request by staffordsmith83**
### What this PR does

Fixes [#156](https://github.com/TerriaJS/product/issues/156)

This PR implements COG support in Cesium Mode.

### Whats been done

- changes to zoom in Cesium.ts for COGs
- changes to zoom in Leaflet. Previously the code tried to access `east` and `west` properties of `Rectangle` but this may be undefined with COGs in 2D mode, so this needs to be handled so zoomTo button doesnt error in 2D mode. Please help me check that this doesnt break anything.
    - `      // Ensure extent is defined before accessing its properties
      if (isDefined(extent)) {`
- Added CogCatalogItem. This is greatly reduced in complexity since moving back to 3D only support.

### Remaining task
There is a separate draft PR to implement a shortReport warning about non-native projection COGs: https://github.com/TerriaJS/terriajs/pull/7226

It is incomplete, but the strategy to finish it is in the description of the PR.

### A note on Reprojecting on the Fly
We had intended to only support COGs prepared in 3857 and 4326 as reprojecting in the browser dynamically is not trivial and was poorly implemented. However the reprojection mechanism used by TIFFImageryProvider has recently been rewritten and is now good enough to be included. It is flagged as expermental, and there is a console warning of 
`[Experimental] Reprojection EPSG:xxxxx` . We should be happy that we have any reprojection on the fly.

Not all projections work. I have identified some for the Australian context that do not:
- EPSG:9473 does not. This throws a warning, as it is not included in the proj4-fully-loaded library. This might be an easy fix with a PR to `proj4-fully-loaded` or a return to the async method of supplying projections to TIFFImageryProvider
- EPSG:3577 does not. This is GDA94 Australian Albers, and is used for some Australian datasets that cover all Australia as provided by DEA. See the 'Unsupported' group of the test data in the testing sharelink. These datasets load without errors, but are not visible on the map.

This PR also relies on https://www.npmjs.com/package/terriajs-tiff-imagery-provider
I have forked this from https://github.com/hongfaqiu/TIFFImageryProvider
And made a new repo at https://github.com/TerriaJS/TIFFImageryProvider

### Test me

http://ci.terria.io/cog-support-cesium-only-1/#clean&https://gist.githubusercontent.com/staffordsmith83/9a6ba12479100a810dadcf0119242913/raw/8c73e5053c4d26180784281840a9c235eface61f/cogs_three_groups.json

There are three groups of COGs in the catalog.
The first group all should work. There are a variety of projections, and tests for feature picking disabled. Also, true color images and single band images.

The second group, none should work.
There are some unsupported projections here, and some COG sources with unsupported photometric interpretations.

The third group is for visually testing that traits make it through.
- credit should be shown in Basemaps link at bottom of map window:
- 
image

A few things like tile size and cache time are difficult to test visually but appear to be getting passed. These could be added into the unit tests if you think necessary.

### Pixel shifting problem

Here is a collection of coarse gridded scientific data. CMIP6 climate projections in EPSG:4326.
I have added one to the testlink catalog gist.

https://nex-gddp-cmip6-cog.s3.us-west-2.amazonaws.com/index.html

### Checklist

- [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
- [ ] I've updated relevant documentation in `doc/`.
- [x] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/Models/Terria.ts [hunk 1]
```diff
@@ -159,7 +159,7 @@ export interface ConfigParameters {
   /**
    * URL of Proj4 projection lookup service (part of TerriaJS-Server).
    */
-  proj4ServiceBaseUrl?: string;
+  proj4ServiceBaseUrl: string;
```
Comment (nf-s) [inline lib/Models/Terria.ts]: I think this should stay optional here - if people want to disable it

File: lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  })
+  nodata?: number;
+
+  @primitiveTrait({
+    type: "boolean",
+    name: "Convert to RGB",
+    description: "Try to render multi band cog to RGB, priority 1"
+  })
+  convertToRGB?: boolean;
+}
```
Comment (nf-s) [inline lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts]: Feel free to put this in `lib/Traits/TraitsClasses/CogCatalogItemTraits.ts` if you want to keep everything in one file

File: package.json [hunk 3]
```diff
@@ -118,6 +118,7 @@
     "flexsearch": "0.7.21",
     "fs-extra": "^7.0.1",
     "geojson-vt": "^3.2.1",
+    "geotiff": "^2.1.3",
```
Comment (nf-s) [inline package.json]: Can this be removed? As it isn't directly used in TerriaJS

File: lib/Models/Catalog/CatalogItems/CogCatalogItem.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  defaultAlpha = undefined as any;
+  defaultBrightness = undefined as any;
+  defaultContrast = undefined as any;
+  defaultGamma = undefined as any;
+  defaultHue = undefined as any;
+  defaultSaturation = undefined as any;
+  defaultMagnificationFilter = undefined as any;
+  defaultMinificationFilter = undefined as any;
+  proxy = undefined as any;
+  tileDiscardPolicy = undefined as any;
```
Comment (nf-s) [inline lib/Models/Catalog/CatalogItems/CogCatalogItem.ts]: Because `TIFFImageryProvider` doesn't extend cesium's `ImageryProvider` I don't think these options are doing anything.

So I think we should remove them, but also change the magnification/minification filter to nearest to get rid of that weird blur. It also reduces artifacts between tiles

```suggestion
  _defaultMagnificationFilter = TextureMagnificationFilter.NEAREST;
  _defaultMinificationFilter = TextureMinificationFilter.NEAREST;
```

**Review by nf-s — COMMENTED** [review]
Awesome stuff, thanks @staffordsmith83 !

I haven't finished my review yet - but thought I should post what I have

Did we end up deciding how to handle projections that aren't cesium native (wgs84 and webmercator)?

If are going to allow reprojection, I think we should add a "short report" so there's a message in the workbench. As it looks like there are some reprojection issues

image

- [Reprojection example - epsg:32756](http://ci.terria.io/cog-support-cesium-only-1/#clean&https://gist.githubusercontent.com/staffordsmith83/9a6ba12479100a810dadcf0119242913/raw/d1c0fdf888e8477be8f044721b7cd1e2ec33c907/cogs_three_groups.json&start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22%2F%2FSupported+Datasets%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22group%22%7D%2C%22%2F%2FSupported+Datasets%2FCOG+Test+Whian+Whian%22%3A%7B%22opacity%22%3A1%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%2FSupported+Datasets%22%5D%2C%22type%22%3A%22cog%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%22%2F%2FSupported+Datasets%2FCOG+Test+Whian+Whian%22%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A153.41161478960097%2C%22south%22%3A-28.061298397857612%2C%22east%22%3A153.42650335865278%2C%22north%22%3A-28.042438926989824%2C%22position%22%3A%7B%22x%22%3A-5039055.0838929955%2C%22y%22%3A2521274.6258177627%2C%22z%22%3A-2982430.527881199%7D%2C%22direction%22%3A%7B%22x%22%3A0.7892395196302652%2C%22y%22%3A-0.3948933959656323%2C%22z%22%3A0.47027671266661586%7D%2C%22up%22%3A%7B%22x%22%3A-0.4205699368373914%2C%22y%22%3A0.21043078364419832%2C%22z%22%3A0.8825190159557363%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22viewerMode%22%3A%223dSmooth%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.4999%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Atrue%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%5D%7D%5D%7D)
- [uluru reprojection example](http://ci.terria.io/cog-support-cesium-only-1/#clean&https://gist.githubusercontent.com/staffordsmith83/9a6ba12479100a810dadcf0119242913/raw/d1c0fdf888e8477be8f044721b7cd1e2ec33c907/cogs_three_groups.json&start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22%2F%2FSupported+Datasets%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22group%22%7D%2C%22%2F%2FSupported+Datasets%2FCOG+Test+Whian+Whian%22%3A%7B%22opacity%22%3A1%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%2FSupported+Datasets%22%5D%2C%22type%22%3A%22cog%22%7D%2C%22%2F%2FSupported+Datasets%2FUSGS+Elevation%22%3A%7B%22opacity%22%3A1%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%2FSupported+Datasets%22%5D%2C%22type%22%3A%22cog%22%7D%2C%22%2F%2FSupported+Datasets%2FCOG+Test+Uluru%281%29%22%3A%7B%22knownContainerUniqueIds%22%3A%5B%22%2F%2FSupported+Datasets%22%5D%2C%22type%22%3A%22cog%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%22%2F%2FSupported+Datasets%2FCOG+Test+Uluru%281%29%22%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A130.5873113853551%2C%22south%22%3A-25.81071236732281%2C%22east%22%3A130.59992099929624%2C%22north%22%3A-25.794414151922386%2C%22position%22%3A%7B%22x%22%3A-3739698.691339093%2C%22y%22%3A4364165.316830322%2C%22z%22%3A-2760066.4969194573%7D%2C%22direction%22%3A%7B%22x%22%3A0.585815228580093%2C%22y%22%3A-0.683636494180984%2C%22z%22%3A0.43527194004160996%7D%2C%22up%22%3A%7B%22x%22%3A-0.2832269316883932%2C%22y%22%3A0.33052105372270196%2C%22z%22%3A0.9002990271084452%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22viewerMode%22%3A%223dSmooth%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.5%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Atrue%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%5D%7D%5D%7D)

I actually think we should disable reprojection until we fully understand how it works - and can validate that it is accurate enough for visualisation purposes - I think it is doing some weird stuff on the edges of tiles - especially when using nearest pixel interpolation.

There are also some tiling and pixel interpolation issues with the 4326 COG

- [open aerial map example](http://ci.terria.io/cog-support-cesium-only-1/#clean&https://gist.githubusercontent.com/staffordsmith83/9a6ba12479100a810dadcf0119242913/raw/d1c0fdf888e8477be8f044721b7cd1e2ec33c907/cogs_three_groups.json&start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22%2F%2FSupported+Datasets%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22group%22%7D%2C%22%2F%2FSupported+Datasets%2FDrone+image+from+Open+Aerial+Map%22%3A%7B%22knownContainerUniqueIds%22%3A%5B%22%2F%2FSupported+Datasets%22%5D%2C%22type%22%3A%22cog%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%22%2F%2FSupported+Datasets%2FDrone+image+from+Open+Aerial+Map%22%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A-90.8706279985145%2C%22south%22%3A14.589539101881737%2C%22east%22%3A-90.8702281108131%2C%22north%22%3A14.590095198398252%2C%22position%22%3A%7B%22x%22%3A-93788.20540361805%2C%22y%22%3A-6173117.815926865%2C%22z%22%3A1596232.9717276145%7D%2C%22direction%22%3A%7B%22x%22%3A0.014701393256942266%2C%22y%22%3A0.9676422770093419%2C%22z%22%3A-0.2518973854181122%7D%2C%22up%22%3A%7B%22x%22%3A0.0038266364342477723%2C%22y%22%3A0.25186831804353843%2C%22z%22%3A0.9677539497307769%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22viewerMode%22%3A%223dSmooth%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.4999%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Atrue%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%5D%7D%5D%7D)

This example gets much better when the `TextureMagnificationFilter` is set to `Nearest` instead of the default `Linear` value - I think we should change the default to `Nearest` (see comment below)

## On texture filters + pixel interpolation

### Reprojection example

Screenshot 2024-06-25 at 5 02 57 PM

### 4326 example

Screenshot 2024-06-25 at 5 25 38 PM

Screenshot 2024-06-25 at 5 44 00 PM

I'm a bit confused why the "linear pixel interpolation with nearest filter" looks so weird - why are the pixels that shape?

This is what the data looks like in QGIS (with web mercator projection)

image

I would expect a 4326 COG to look basically perfect in cesium

### Web mercator example

I need to find an example to use here

**Comment by staffordsmith83** [comment]
Yeah there are some funny issues there arent there unfortunately. It appears that the QGIS image looks identical to the "Nearest Pixel + Nearest Filter" example you provided, what do you think?

I am open to either disabling reprojection, or to add short reports. But I feel we should allow reprojection, and just put a short report that it is experimental. Certainly I can think of uses where I dont mind some edge effects but would be great to use terria to show a preview of a cog I have in a different projection. If I need the absolute best visualisation I can get the link for the resource and download it or whatever.

I suppose its up to people who assemble the catalogs - if you want it to be rock solid, reproject to 4326 (hopefully!)
Shall I reproject some of those examples to 3857 to compare the same datasets perhaps?

Thanks for taking such a close look, this is awesome to have other eyes on it. More tomorrow!

File: lib/Models/Terria.ts (hunk 1 above)
Comment (staffordsmith83) [reply lib/Models/Terria.ts]: OK will do. That was from an earlier change Im sure it can go back to optional no problem

File: lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts (hunk 2 above)
Comment (staffordsmith83) [reply lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts]: OK can do that. Its going to grow if I have time for second traits PR though

File: package.json (hunk 3 above)
Comment (staffordsmith83) [reply package.json]: yes, was from previous code when using TIFFImageryProvider in third part folder sorry

File: lib/Models/Catalog/CatalogItems/CogCatalogItem.ts (hunk 4 above)
Comment (staffordsmith83) [reply lib/Models/Catalog/CatalogItems/CogCatalogItem.ts]: Great find! Yes, once again these were to fix something in one of my earlier iterations, hopefully no longer needed.

File: lib/Models/Terria.ts (hunk 1 above)
Comment (staffordsmith83) [reply lib/Models/Terria.ts]: done

File: lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts (hunk 2 above)
Comment (staffordsmith83) [reply lib/Traits/TraitsClasses/CogRenderOptionsTraits.ts]: done

File: lib/Models/Catalog/CatalogItems/CogCatalogItem.ts (hunk 4 above)
Comment (staffordsmith83) [reply lib/Models/Catalog/CatalogItems/CogCatalogItem.ts]: done

File: package.json (hunk 3 above)
Comment (staffordsmith83) [reply package.json]: done

**Comment by staffordsmith83** [comment]
Just going to implement short report now. We can disable all reprojection later if decided.

**Review by nf-s — CHANGES_REQUESTED** [review]
The "nearest pixel + nearest filter" 4326 example is much better than "nearest pixel + linear filter" - but it is still doing weird things with the pixel sizes. The no data values are also quite different from the actual data.

image

Any COG in 4326 or 3857 _should_ be able to be visualised without these kinds of artefacts in cesium. What also concerns me is that when you change `resampleMethod` to `linear` you get these really narrow pixels, which should be square(ish). I think there are bugs in how `TIFFImageryProvider` calculates the number of pixels (and how they are arranged) in each tile/overview.

I don't mind if the data isn't perfectly represented at lower zoom levels/lower LOD - but as you zoom in it should get more accurate, and then at some point you should be able to see actual pixel values. This may not be that important for imagery data, but I think it is more important for scientific gridded data. For example, if we were to bring in climate projection data, which is quite coarse spatially, I'm worried that the visualisation may be misrepresenting the data.

I think there may be some fundamental issues with `TIFFImageryProvider` here. I was aware of possible issues with how it approaches reprojection - but I wasn't aware that it has issues with CRSs that are native to cesium. 

I apologise if this seems overly critical. 

As we add more "raw" data formats to Terria - especially a new class of format (as this is the first raster format) - I think it is very important that we can be certain that data is not being misrepresented. At this stage I don't think we should merge this in until we have identified the issues in `TIFFImageryProvider`. In my opinion, I think the minimum requirements for Geotiff/COG support in Terria is support for 4326 and 3857, and when you zoom in you should be able to see the actual pixel values with reasonable accuracy.

@na9da and @steve9164 what do you think?

**Comment by staffordsmith83** [comment]
I can find some netCDF CMIP6 data or similar climate data to add and see how that looks too.
No worries I think its important to preserve the quality of the Terria product, if you guys dont think the way TIFFImageryProvider works is rigorous enough then I dont mind. I agree misrepresentation of data is a dangerous thing.

**Comment by staffordsmith83** [comment]
@nf-s Ill add some gridded climate data examples from here: https://registry.opendata.aws/nex-gddp-cmip6/.

UPDATE: Added one to the 'Supported' group in the test sharelink gist. It looks bad.
Data should be square pixels:
image

It may be worth testing the [COG Url](https://nex-gddp-cmip6-cog.s3-us-west-2.amazonaws.com/daily/GFDL-CM4/ssp585/r1i1p1f1/pr/pr_day_GFDL-CM4_ssp585_r1i1p1f1_gr1_2015_01_25.tif) in the TIFFImageryProvider online demo: https://tiff-imagery-provider.opendde.com/

**Comment by staffordsmith83** [comment]
The dataset should look like this: 
image

**Comment by staffordsmith83** [comment]
Here is a separate draft PR to implement a shortReport warning about non-native projection COGs: https://github.com/TerriaJS/terriajs/pull/7226

It is incomplete, but the strategy to finish it is in the description of the PR.

**Comment by staffordsmith83** [comment]
Regarding resampling error: Hongfaqiu is aware of the issue and is going to try to fix it next week. With a bit of luck after that a merge from upstream and COGs will be good to go! https://github.com/hongfaqiu/TIFFImageryProvider/issues/30#issuecomment-2193947286

**Comment by imakihi** [comment]
@staffordsmith83
Thank you for working on this! I checked if the value picking issue I described in https://github.com/TerriaJS/terriajs/pull/6833 was solved. It looks like the value picking on the 3D mode working correctly!!! Great!!

**Comment by staffordsmith83** [comment]
Dear @imakihi thats great! Im no longer working on the Terria team, so Im not sure when they plan to merge this, @nf-s whats the plans with COGs? I could try to find some time over the summer to help get it done?

**Comment by imakihi** [comment]
Hi @staffordsmith83 
I didn't know that you left the Terria team. Thank you for starting this great feature and worked on it.
I think this COG capability benefits many Terria users.
Thank you!

**Comment by staffordsmith83** [comment]
> Hi @staffordsmith83 I didn't know that you left the Terria team. Thank you for starting this great feature and worked on it. I think this COG capability benefits many Terria users. Thank you!

My pleasure Hiroo! Hope to meet you some day at a conference or something :-)

**Comment by nf-s** [comment]
> Dear @imakihi thats great! Im no longer working on the Terria team, so Im not sure when they plan to merge this, @nf-s whats the plans with COGs? I could try to find some time over the summer to help get it done?

We are in a bit of a frenzy at the moment, and I'm about to be off for 2 weeks, but it would be awesome to get this merged in soon. Thanks for offering your time @staffordsmith83! That is very kind of you. I'll see if someone in the team has time to take a look first to save you the effort - I'll let you know if I can find someone to look at it over the next 2 weeks 🙂

**Comment by imakihi** [comment]
@nf-s Thank so much you. It would be super super cool if Terria users could use COG data.

**Comment by staffordsmith83** [comment]
Thanks @nf-s yes let me know, Im crazy flat out but I would just love to see this get finished, lets make it happen together!

**Comment by na9da** [comment]
@staffordsmith83 - Thanks for offering to help. I'll get some time next week to have a look. I'll need to get familiar with COG stuff though. Will ping here if I have any questions :)

**Comment by staffordsmith83** [comment]
@na9da OK Ill keep an eye on this ticket!

**Comment by imakihi** [comment]
@na9da @nf-s 
I was wondering if you had time to think about this COG support...

**Comment by na9da** [comment]
@imakihi - sorry i haven't had a chance to look at it yet. But hopefully soon.

**Comment by na9da** [comment]
Hey @staffordsmith83 @nf-s - I have updated the PR with the following main changes:

- [Upgraded](https://github.com/TerriaJS/TIFFImageryProvider/pull/1) terriajs-tiff-imagery-provider to match latest upstream `main`
  - `linear` mode has been replaced with a better `bilinear` mode which looks great, so I have made it the default.
     Here's a comparison of nearest (left) & bilinear (right) modes ([CI test link](http://ci.terria.io/cog-support-cesium-only-1/#share=s-t5i1wQbevr0dGRXNqUgTfjMNHcl&clean&https://gist.githubusercontent.com/na9da/18ce40ab4d93a722932c3efe5c947139/raw/b06ed050745f57ba9b081e7700e6504ea6719a9b/cogs.json))
  - minimized custom changes so that merges from upstream becomes easier (turned of prettier etc)
  - changes [documented here](https://github.com/TerriaJS/TIFFImageryProvider/blob/main/packages/TIFFImageryProvider/CHANGES-FOR-TERRIA.md)
- CogCatalogItem changes
  - switched imagery provider loading to use the `fromUrl` API (the `url` option is deprecated in upstream).
  - added short report when using non-standard projections
 
Pending: update test cases

**Comment by na9da** [comment]
@nf-s new test url - http://ci.terria.io/cog-support-cesium-only-1/#share=s-t5i1wQbevr0dGRXNqUgTfjMNHcl&clean&https://gist.githubusercontent.com/na9da/18ce40ab4d93a722932c3efe5c947139/raw/b06ed050745f57ba9b081e7700e6504ea6719a9b/cogs.json

**Review by nf-s — CHANGES_REQUESTED** [review]
Hi @na9da 

Looks great @na9da ! I checked the 4326 and 3857 examples and compared with QGIS, and the output looks good! 

I think the default resampling method should be "nearest" to most closely represent the data - while it might make sense to apply linear/billinear transformation to imagery by default - a lot, if not most, raster data is not imagery, it is scientific data, where the pixel bounds are quite important. There are also weird artifacts along tile borders with linear/billinear resampling.

Is it possible to up the "quality" of the COGs? It looks like the imagery provider is always a step behind, so you can always see pixels - even with the quality slider set to "max quality".

For example - here I can clearly see the pixels, but the TIFF is has quite a few more level-of-details to load

- http://ci.terria.io/cog-support-cesium-only-1/#share=s-mY9ifiWSeLGF9ZCQVunRv0iIH7v&clean&https://gist.githubusercontent.com/na9da/18ce40ab4d93a722932c3efe5c947139/raw/b06ed050745f57ba9b081e7700e6504ea6719a9b/cogs.json

image

One other small comment, when I paste in a TIFF into "Add Web Data" without specifying "Cloud Optimized GeoTIFF", there is a error - it would be good to have support for `.tif`, `.tiff`, `.geotiff`

Thanks!

**Comment by nf-s** [comment]
Also did a quick test to see if alignment is correct - and it looks to be good!

image

**Comment by na9da** [comment]
On the quality, i had a look - the particular opendrone dataset has 8 images in the TIFF pyramid. The imagery provider seems to be correctly picking and cycling through them as Cesium requests new zoom levels.

Not sure exactly whats causing the pixelized look. In 3D mode, there might be different active zoom levels (see screenshot below), which could be one reason.
image

However, I switched the map to 2d mode, which resulted in a single zoom level for all tiles, but the pixelization didn't improve much.

The other thought is, the COG has 8 images, but does that map linearly with zoom levels, i.e at zoom=5, is it correct to use the 5th image in the COG, or is there some other math to figure out which zoom level to use.

Also, not sure if there is any other stretching happening from adjusting for image shape (aspect ratio related).

The other possibility, is some processing happening in cesium that makes the images look less sharp. I tried changing the cesium imagery layer to use "NEAREST" sampling ([as suggested here](https://community.cesium.com/t/blurry-imagery-faa-vfr-tiles/31540/3)), which makes it slightly less blurry 
Edit: I see you already explored this in a previous comment

image
(left with NEAREST filter slightly sharper vs right with default linear)

I also compared it with openlayers, which looks super crisp. No idea how to get anywhere close to that.

image
(openlayers - no pixelation at all)

TLDR; not much clue where that pixelation is coming from.

**Comment by staffordsmith83** [comment]
Great to see the upstream changes have improved this! Ill try to find some some for a closer look next week. Thanks @na9da !

**Comment by nf-s** [comment]
> On the quality, i had a look - the particular opendrone dataset has 8 images in the TIFF pyramid. The imagery provider seems to be correctly picking and cycling through them as Cesium requests new zoom levels.
> 
> Not sure exactly whats causing the pixelized look. In 3D mode, there might be different active zoom levels (see screenshot below), which could be one reason. image
> 
> However, I switched the map to 2d mode, which resulted in a single zoom level for all tiles, but the pixelization didn't improve much.
> 
> The other thought is, the COG has 8 images, but does that map linearly with zoom levels, i.e at zoom=5, is it correct to use the 5th image in the COG, or is there some other math to figure out which zoom level to use.
> 
> Also, not sure if there is any other stretching happening from adjusting for image shape (aspect ratio related).
> 
> The other possibility, is some processing happening in cesium that makes the images look less sharp. I tried changing the cesium imagery layer to use "NEAREST" sampling ([as suggested here](https://community.cesium.com/t/blurry-imagery-faa-vfr-tiles/31540/3)), which makes it slightly less blurry Edit: I see you already explored this in a previous comment
> 
> image (left with NEAREST filter slightly sharper vs right with default linear)
> 
> I also compared it with openlayers, which looks super crisp. No idea how to get anywhere close to that.
> 
> image (openlayers - no pixelation at all)
> 
> TLDR; not much clue where that pixelation is coming from.

All good, thanks for checking @na9da !
The quality issue is just a nice-to-have for me. We can improve it later. It could be something in TIFFImageryProvider causing the issue.

Thanks a lot @staffordsmith83 and @na9da !
We're so close to merging this in 😄 !

**Comment by na9da** [comment]
@nf-s thanks for the review.

I have now fixed some additional CI errors (related to a change in @types/leaflet).

Also made [this change](https://github.com/TerriaJS/terriajs/pull/7209/commits/afa94ef3b277c536bb6e14111dc138a06786777c) to free the imageryProvider when not used. The imageryProvider creates a bunch of webworkers. This change ensures that they get freed when the item is removed from the workbench. Ideally, we would have a `destroy()` life-cycle method for catalog items that would make this more explicit instead of relying on `onBecomeUnobserved`. Maybe we should add it when we deal with https://github.com/TerriaJS/terriajs/issues/4816

**Review by nf-s — APPROVED** [review]
LGTM thanks a lot @staffordsmith83 and @na9da !

**Comment by staffordsmith83** [comment]
Awesome, its in! Thanks @na9da and @nf-s

**Comment by imakihi** [comment]
Super cool. @na9da @nf-s @staffordsmith83
